doc/piaware.1
.\"
.\" Copyright (c) 2014 FlightAware LLC
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\"
.Dd September 24, 2014
.Dt PIAWARE 1
.Os
.Sh NAME
.Nm piaware
.Nd forward ADS-B messages to FlightAware
.Sh SYNOPSIS
.Nm piaware
.Bk -words
.Op Fl p Ar pid_file
.Op Fl debug
.Op Fl help
.Ek
.Sh DESCRIPTION
.Nm
forwards ADS-B messages to FlightAware over an encrypted, compressed
Internet connection.
.Pp
.Nm
established an encrypted, compressed TLS connection with FligthAware and logs 
in using a FlightAware user name and password.
At the same time it tries to
connect to a program on the local computer providing ADS-B messages in 
"beast" binary format, typically \fBdump1090\fR.
.Pp
It then filters and coalesces those messages and forwards them over the
TLS connection where they contribute to the accuracy of FlightAware's
flight tracking.
.Pp
(People who forward ADS-B messages to FlightAware automatically qualify for
a free upgrade to an Enterprise Account.  For more information please visit
http://flightaware.com/adsb/piaware/)
.Pp
On startup piaware attempts to connect to port 10001 to obtain the filtered,
coalesced ADS-B messages.  If the machine is running the FlightAware-modified
version of dump1090 (https://github.com/flightaware/dump1090_mr) then it
will obtain those messages directly.
.Pp
If the site is not running FA-modified dump1090 or is running a different
ADS-B provider program such as modesmixer then piaware will start a helper
program called faup1090 to translate between beast format and FlightAware
format.
.Pp
If installed from the Debian package, piaware will automatically start up at system boot time.
.Pp
Since other programs such as dump1090 may not have started yet, piaware will wait until something is there to provide ADS-B data.  Also piaware will automatically reconnect if that program stops and is started again and it will also periodically try to reconnect to FlightAware in the event that contact is lost.
.Pp
Options:
.Bl -tag -width Ds
.It Fl p Ar pid_file
Specifies the path of a file to store the process ID of the program so it can ensure that a second copy isn't accidentally started and whatnot.  This is used by the /etc/init.d/piaware script and would not normally be used from the command line.
.It Fl debug
If this option is selected
.Nm
will log to stdout rather than to /tmp/piaware.out.
.It Fl help
Emit a brief summary of piaware usage options.
.El
.Sh EXAMPLES
Run piaware from the command line with debug enabled:
.Pp
.Dl $ piaware -debug
.Sh SEE ALSO
.Xr piaware-config 1 ,
.Xr piaware-status 1 ,
.Xr faup1090


package/piaware.tcl
#
# piaware package - Copyright (C 2014 FlightAware LLC
#
# Berkeley license
#

set piawarePidFile /var/run/piaware.pid
set piawareConfigFile /etc/piaware

#
# load_piaware_config - load the piaware config file.  don't stop if it
#  doesn't exist
#
# return 1 if it loaded cleanly, 0 if it had a problem or didn't exist
#
proc load_piaware_config {} {
    if {[catch [list uplevel #0 source $::piawareConfigFile]] == 1} {
	return 0
    }
    return 1
}

#
# query_piaware_pkg - Figure out piaware package name and version
#
# Find a package that has piaware in the name.  There will about for sure
# be only one.
#
# Parse out the name and version into passed-in variables and return
#  1 if successful or 0 if unsuccessful
#
proc query_piaware_pkg {_packageName _packageVersion} {
    upvar $_packageName packageName $_packageVersion packageVersion

    set fp [open "|dpkg-query --show *piaware* 2>/dev/null"]
    gets $fp line

    if {[catch {close $fp}] == 1} {
	return 0
    }

    if {![regexp {([^\t]*)\t(.*)} $line dummy packageName packageVersion]} {
	return 0
    }

    return 1
}

#
# load_piaware_config_and_stuff - invoke load_piaware_config and if it
#   doesn't define imageType then see if the piaware package is installed
#   and if it is then set imageType to package
#
proc load_piaware_config_and_stuff {} {
    load_piaware_config

    if {![info exists ::imageType]} {
	if {[query_piaware_pkg packageName packageVersion]} {
	    set ::imageType "${packageName}_package"
	}
    }
}

# is_pid_running - return 1 if the specified process ID is running, else 0
#
proc is_pid_running {pid} {
    if {[catch {kill -0 $pid} catchResult] == 1} {
	switch [lindex $::errorCode 1] {
	    "EPERM" {
		return 1
	    }

	    "ESRCH" {
		return 0
	    }

	    default {
		error "is_pid_running unexpectedly got '$catchResult' $::errorCode"
	    }
	}
    }
    return 1
}

#
# is_process_running - return 1 if at least one process named "name" is
#  running, else 0
#
proc is_process_running {name} {
    set fp [open "|ps -C $name -o pid="]
    while {[gets $fp line] >= 0} {
	set pid [string trim $line]
	if {[is_pid_running $pid]} {
	    catch {close $fp}
	    return 1
	}
    }
    catch {close $fp}
    return 0
}

#
# is_piaware_running - find out if piaware is running by checking its pid
#  file
#
proc is_piaware_running {} {
    if {[catch {set fp [open $::piawarePidFile]}] == 1} {
	return 0
    }

    gets $fp pid
    close $fp

    if {![string is integer -strict $pid]} {
	return 0
    }

    return [is_piaware_running]
}

#
# test_port_for_traffic - connect to a port and
#  see if we can read a byte before a timeout expires.
#
# invokes the callback with a 0 for no data received or a 1 for data recv'd
#
proc test_port_for_traffic {port callback {waitSeconds 60}} {
    if {[catch {set sock [socket localhost $port]} catchResult] == 1} {
	puts "got '$catchResult'"
	{*}$callback 0
	return
    }
    fconfigure $sock -buffering none \
	-translation binary \
	-encoding binary

    set timer [after [expr {$waitSeconds * 1000}] [list test_port_callback "" $sock 0 $callback]]
    fileevent $sock readable [list test_port_callback $timer $sock 1 $callback]
}

#
# test_port_callback - routine used by test_port_for_traffic to cancel
#  the timer and close the socket and invoke the callback
#
proc test_port_callback {timer sock status callback} {
    if {$timer != ""} {
	catch {after cancel $timer}
    }
    catch {close $sock}
    {*}$callback $status
}


proc process_netstat_socket_line {line} {
    lassign $line proto recvq sendq localAddress foreignAddress state pidProg
    lassign [split $pidProg "/"] pid prog

    if {$localAddress == "*:30005" && $state == "LISTEN"} {
	set ::netstatus(program_30005) $prog
	set ::netstatus(status_30005) 1
    }

    if {$localAddress == "*:10001" && $state == "LISTEN"} {
	set ::netstatus(program_10001) $prog
	set ::netstatus(status_10001) 1
    }


    switch $prog {
	"faup1090" {
	    if {$foreignAddress == "localhost:30005" && $state == "ESTABLISHED"} {
		set ::netstatus(faup1090_30005) 1
	    }
	}

	"piaware" {
	    set ::running(piaware) 1
	    if {$foreignAddress == "localhost:10001" && $state == "ESTABLISHED"} {
		set ::netstatus(piaware_10001) 1
	    }

	    if {$foreignAddress == "eyes.flightaware.com:1200" && $state == "ESTABLISHED"} {
		set ::netstatus(piaware_1200) 1
	    }
	}
    }
}

#
# inspect_sockets_with_netstat - run netstat and make a report
#
proc inspect_sockets_with_netstat {} {
    set ::running(dump1090) 0
    set ::running(faup1090) 0
    set ::running(piaware) 0
    set ::netstatus(status_30005) 0
    set ::netstatus(status_10001) 0
    set ::netstatus(faup1090_30005) 0
    set ::netstatus(piaware_10001) 0
    set ::netstatus(piaware_1200) 0

    set fp [open "|netstat --program --protocol=inet --tcp --wide --all"]
    # discard two header lines
    gets $fp
    gets $fp
    while {[gets $fp line] >= 0} {
	process_netstat_socket_line $line
    }
    close $fp
}

#
# subst_is_or_is_not - substitute "is" or "is not" into a %s in string
#  based on if value is true or false
#
proc subst_is_or_is_not {string value} {
    if {$value} {
	set value "is"
    } else {
	set value "is NOT"
    }

    return [format $string $value]
}

#
# netstat_report - parse netstat output and report
#
proc netstat_report {} {
    inspect_sockets_with_netstat

    foreach port "30005 10001" {
	set statusvar "status_$port"
	set programvar "program_$port"

	if {!$::netstatus($statusvar)} {
	    puts "no program appears to be listening for connections on port $port."
	} else {
	    puts "$::netstatus($programvar) is listening for connections on port $port."
	}
    }

    if {$::netstatus(faup1090_30005)} {
	puts "faup1090 is connected to port 30005"
    }

    puts "[subst_is_or_is_not "piaware %s connected to port 10001." $::netstatus(piaware_10001)]"

    puts "[subst_is_or_is_not "piaware %s connected to FlightAware." $::netstatus(piaware_1200)]"
}

#
# reap_any_dead_children - wait without delay until we reap no children
#
proc reap_any_dead_children {} {
    # try to reap any dead children
    while {true} {
	if {[catch {wait -nohang} catchResult] == 1} {
	    # got an error, probably no children
	    return
	}

	# didn't get an error
	if {$catchResult == ""} {
	    # and it didn't return anything, we have extant children but
	    # none have exited (or died from a signal) right now
	    return
	}

	#logger "reaped child $catchResult"

	lassign $catchResult pid type code

	switch $type {
	    "EXIT" {
		switch $code {
		    default {
			logger "the system told us that process $pid exited due to some general error"
		    }
		    98 {
			logger "the system confirmed that process $pid exited.  the exit status of $code tells us that faup1090 couldn't open the listening port because something else already has it open"
		    }

		    0 {
			logger "the system told us that process $pid exited cleanly"
		    }
		}
		logger "the system confirmed that process $pid exited with an exit status of $code"
	    }

	    "SIG" {
		if {$code == "SIGHUP"} {
		    logger "the system confirmed that process $pid exited after receiving a hangup signal"
		} else {
		    logger "this is a little unexpected: the system told us that process $pid exited after receiving a $code signal"
		}
	    }

	    default {
		logger "the system told us one of our child processes exited but i didn't understand what it said: $catchResult"
	    }
	}
    }
}

#
# get_local_device_ip_address - figure out the specified device's IP address
#
# note - does not cache, returns empty string if the machine doesn't
#  have one
#
proc get_local_device_ip_address {dev} {
    set fp [open "|ip address show dev $dev"]
    while {[gets $fp line] >= 0} {
        if {[regexp {inet ([^/]*)} $line dummy ip]} {
            catch {close $fp}
            return $ip
        }
    }
    # didn't find it, command might not have worked, make sure trying to
    # close it doesn't cause a traceback
    catch {close $fp}
    if {$dev == "eth0"} {
	warn_once "failed to get mac address for this computer. piaware will not work properly without it! are you running piaware on something other than a raspberry pi? piaware may need to be modified"
    }
    return ""
}

#
# get_local_ethernet_ip_address - figure out the ethernet port's IP address
#
proc get_local_ethernet_ip_address {} {
    return [get_local_device_ip_address eth0]
}

#
# get_default_gateway_interface_and_ip - assign the default gateway and 
#  interface to the passed-in variables and return 1 if successful in
# determining, else return 0
#
proc get_default_gateway_interface_and_ip {_gateway _iface _ip} {
    upvar $_gateway gateway $_iface iface $_ip ip

    set fp [open "|netstat -rn"]
    gets $fp
    gets $fp

    while {[gets $fp line] >= 0} {
	if {[catch {lassign $line dest gateway mask flags mss window irtt iface}] == 1} {
	    continue
	}
	if {$dest == "0.0.0.0"} {
	    close $fp
	    set ip [get_local_device_ip_address $iface]
	    return 1
	}
    }
    close $fp
    return 0
}

#
# warn_once - issue a warning message but only once
#
proc warn_once {message args} {
    if {[info exists ::warnOnceWarnings($message)]} {
	return
    }
    set ::warnOnceWarnings($message) ""

    logger "WARNING $message"
}

#
# upgrade_raspbian - upgrade raspbian to the latest packages, kernel,
#  libraries, boot files and whatnot
#
proc upgrade_raspbian {} {
    logger "*** attempting to upgrade raspbian to the latest"
    logger "doing apt-get update..."
    set exitStatus [system "apt-get --yes update"]
    if {$exitStatus != 0} {
	logger "got exit status $exitStatus, aborting..."
	return 0
    }

    logger "doing apt-get upgrade..."
    set exitStatus [system "apt-get --yes upgrade"]
    if {$exitStatus != 0} {
	logger "got exit status $exitStatus, aborting..."
	return 0
    }
    logger "doing rpi-update..."
    set exitStatus [system "rpi-update"]
    if {$exitStatus != 0} {
	logger "got exit status $exitStatus, aborting..."
	return 0
    }

    system "/sbin/reboot"
    return 1
}

package provide piaware 1.0
